A29-xds-tls-security.md

```diff
@@ -0,0 +1,446 @@
+XdsServerCredentials is used as described in the example snippets below. In these examples,
+a plaintext or insecure credentials is used as the fallback credentials.
+
+
+#### C++ 
+
+Use of XdsChannelCredentials:
+
+```C++
+    std::shared_ptr<grpc::ChannelCredentials> credentials =
+        grpc::experimental::XdsCredentials(grpc::InsecureChannelCredentials());
+    std::shared_ptr<grpc::Channel> channel = grpc::CreateChannel(target, credentials);
+```
+
+Use of XdsServerCredentials:
+
+```C++
+   grpc::experimental::XdsServerBuilder builder;
+   builder.RegisterService(&service);
+   builder.AddListeningPort(listening_address,
+                                 grpc::experimental::XdsServerCredentials(
+                                     grpc::InsecureServerCredentials()));
+   builder.BuildAndStart()->Wait();
+```
+
+#### Java
+
+Use of XdsChannelCredentials:
+
+```Java
+    ChannelCredentials credentials = XdsChannelCredentials.create(InsecureChannelCredentials.create());
+    ManagedChannel channel = Grpc.newChannelBuilder(target, credentials).build();
+```
+
+Use of XdsServerCredentials:
+
+```Java
+    ServerCredentials credentials = XdsServerCredentials.create(InsecureServerCredentials.create());
+    Server server = XdsServerBuilder.forPort(port, credentials)
+      .addService(new HostnameGreeter(hostname))
+      .build()
+      .start();
+```
+
+#### Go
+
+Use of XdsChannelCredentials:
+
+```Go
+   import (
+           xdscreds "google.golang.org/grpc/credentials/xds"
+   )
+
+   if creds, err := xdscreds.NewClientCredentials(xdscreds.ClientOptions{FallbackCreds: insecure.NewCredentials()}); err != nil {
+           log.Fatalf("Failed to create xDS client credentials: %v", err)
+   }
+   conn, err := grpc.Dial(*target, grpc.WithTransportCredentials(creds))
+```
+
+Use of XdsServerCredentials:
+
+```Go
+   import (
+           xdscreds "google.golang.org/grpc/credentials/xds"
+           "google.golang.org/grpc/xds"
+   )
+
+   if creds, err := xdscreds.NewServerCredentials(xdscreds.ServerOptions{FallbackCreds: insecure.NewCredentials()}); err != nil {
+           log.Fatalf("Failed to create xDS server credentials: %v", err)
+   }
+   server := xds.NewGRPCServer(grpc.Creds(creds))
+```
+
+### xDS Protocol
+
+gRPC uses CDS to configure client side security and LDS for server side security as described below.
+The required fields were added to xDS v3 hence xDS v3 is a pre-requisite for this proposal.
+
+#### CDS for Client Side Security
+
+The CDS policy is the top level LB policy applied to all the connections under that policy.
+[A27:CDS][] describes the flow. The field [`transport_socket`][CL-TS] is used to extract the
+[`UpstreamTlsContext`][UTC] as described [here][CL-TS-comment].
+Note that we don't (currently) support [`transport_socket_matches`][CL-TS-matches].
+The `UpstreamTlsContext` thus obtained is passed down to all the child policies and connections.
```

Comment:
The actual `UpstreamTlsContext` message is not passed down to the child policies. Instead whatever information is of use to us is extracted and that is what is passed down. Now sure if that distinction is useful to highlight here.